LSLib/Granny/ClothUtils.cs

```diff
@@ -0,0 +1,384 @@
+
+            for (int i = 0; i < points.Length; i++)
+            {
+                points[i] = vertices[i].PositionAsArray;
+            }
+
+            // don't use Math.Sqrt and Math.Pow for the metric since this'll be called thousands of times (square distance is fine here)
+            return new KDTree<float, T>(3, points, vertices, (a, b) => (b[0] - a[0]) * (b[0] - a[0]) + (b[1] - a[1]) * (b[1] - a[1]) + (b[2] - a[2]) * (b[2] - a[2]));
+        }
+
+        private static PhysicsVertex[] GetPhysicsClothVertices(ClothMesh mesh)
+        {
+            var markedVertices = GetClothVertices(mesh);
+            var encountered = new HashSet<Vector3>();
+            var physicsVertices = new List<PhysicsVertex>(markedVertices.Count);
+
+            for (int i = 0; i < mesh.Indices.Count / 3; i++)
+            {
+                ClothVertex v1 = mesh.Vertices[mesh.Indices[i * 3]];
+                ClothVertex v2 = mesh.Vertices[mesh.Indices[i * 3 + 1]];
+                ClothVertex v3 = mesh.Vertices[mesh.Indices[i * 3 + 2]];
+
+                // skip the triangle if it has a non-cloth vertex
+                if (!markedVertices.Contains(v1) || !markedVertices.Contains(v2) || !markedVertices.Contains(v3))
+                {
+                    continue;
+                }
+
+                for (int j = 0; j < 3; j++)
+                {
+                    ClothVertex vertex = mesh.Vertices[mesh.Indices[i * 3 + j]];
+
+                    // overlapping vertices should all map to the first encountered
+                    if (encountered.Contains(vertex.Position))
+                    {
+                        continue;
+                    }
+
+                    physicsVertices.Add(new PhysicsVertex((short)physicsVertices.Count, vertex));
+                    encountered.Add(vertex.Position);
+                }
+            }
+
+            return [.. physicsVertices];
+        }
+
+        private static ClothVertex[] GetTargetClothVertices(ClothMesh mesh)
+        {
+            ClothVertex[] clothVertices = [.. GetClothVertices(mesh).OrderBy(v => v.Index)];
+            var packedVertices = new ClothVertex[clothVertices.Length];
+
+            var i = 0;
+
+            for (; i < clothVertices.Length && clothVertices[i].Index < clothVertices.Length; i++)
+            {
+                var vertex = clothVertices[i];
+                packedVertices[vertex.Index] = vertex;
+            }
+
+            var current = 0;
+
+            for (; i < clothVertices.Length; i++)
+            {
+                while (packedVertices[current] != null)
+                {
+                    current++;
+                }
+
+                packedVertices[current++] = clothVertices[i];
+            }
+
+            return packedVertices;
+        }
+
+        private static HashSet<ClothVertex> GetClothVertices(ClothMesh mesh)
+        {
+            var clothVertices = new HashSet<ClothVertex>(mesh.Vertices.Where(v => v.Weight > 0));
```

Comment:
It might be easier to store an `Added` flag in ClothVertex.
This improves things in 3 ways:
1) No need to build a HashSet out of verts
2) `AddOverlappingVertices` and `AddNeighboringVertices` does not need to iterate over the complete set of vertices each time it is called, and re-check each vertex against the HashSet to see if it was added already; instead, if verts were kept in a List, `AddOverlappingVertices` and `AddNeighboringVertices` could just check the vert list starting from the last index they inspected, hence skipping verts whose neighbors and overlaps were already added.
3) `GetTargetClothVertices` is simplified, as the output of `GetClothVertices` already contains a packed list of verts, so there is no need to separately build `packedVertices`

Reply:
You're totally right about the first two points, but I don't think the 3rd one is possible since I need the final number of vertices to build `packedVertices` in the right order. I'm not sure it'd really be easier with an `Added` flag, but I'll see what I can do.